haystack/nodes/preprocessor/preprocessor.py

```diff
@@ -315,72 +418,103 @@ def split(
         if type(document.content) is not str:
             logger.error("Document content is not of type str. Nothing to split.")
             return [document]
-
+            
+        if pre_split_paragraphs and split_by == "passage":
+            logger.error('"pre_split_paragraphs=True" is not compatible with split_by="passage"')
+            return [document]
```

Comment:
Oh, I might have missed those. Yes, in general I think it's very important that the preprocessor either fails or splits the documents. Can you list here the places where the documents are returned whole? Feel also free to fix them right away. Listing them just makes it easier for me to re-review.

Regarding _what_ exactly to do, there doesn't need to be a strict consistency, but some would be nice. For example `raise Exception` is almost never a good idea: let's throw a meaningful exception, something like `ValueError` if a parameter was wrong, `NotImplementedError` if there's a TODO nearby, etc... I'll have a look at the others case by case.

Reply:
I seemingly can't add comments outside of your diff so I'll list them here.

#### 1. 
```python
        if not isinstance(document, Document):
            raise HaystackError("Document must not be of type 'dict' but of type 'Document'.")
```
This one is ok.

#### 2.
```python
        if not split_by:
            return [document]
```
This is not ok :smile: Let's replace it with:
```python
        if not split_by in ["word", "sentence", "passage"]:
            raise ValueError('split_by must be one of: "word", "sentence", "passage", not "%s"', split_by)
```
Apparently passing `split_by==None` was used to disable splitting, according to the docstring. I don't think this feature is ever used, and my guess that's reflected by the typing, so let's update the docs to remove that option.

#### 3.
```python
        if not split_length:
            raise Exception("split_length needs be set when using split_by.")
```
Almost ok: let's use `ValueError` instead of `Exception` (bare `Exception`s are really hard to handle properly in code). `ValueError`s are perfect when the error originated from a wrong parameter value.

#### 4.
```python
        if split_respect_sentence_boundary and split_by != "word":
            raise NotImplementedError("'split_respect_sentence_boundary=True' is only compatible with split_by='word'.")
```
I think a `ValueError` makes more sense here too.

#### 5.
```python
        if type(document.content) is not str:
            logger.error("Document content is not of type str. Nothing to split.")
            return [document]
```
This, for as weird as it looks, is fine. It's necessary to not make the PreProcessor crash when it meets `table` type Documents, which cannot (and should not) be split.

#### 6.
```python
        if pre_split_paragraphs and split_by == "passage":
            raise ValueError('"pre_split_paragraphs=True" is not compatible with split_by="passage"')
```
Good as it is.

#### 7.
```python
        if split_overlap > split_length:
            raise ValueError("split_length must be greater than split_overlap")
```
Good as it is.